File: python/tri_temp.py
#Main Project:
#Take and validate numeric-input, and unit: Units (F/C/K)
#Maximum Temperature included, when converted displays all three and asks to run again.
ABS_ZERO_C = -273.15
ABS_ZERO_F = -459.67
ABS_ZERO_K = 0.0

def get_Temp():
    """Read a numeric temperature from user (re-prompts on non-numeric)"""
    while True:
        try:
            temp = float(input("Enter a Temperature: "))
            #print(temp) #Tester Flag Pass{X} --Notes: Number ran through, attempt characters/string; 
            #Manage ValueError for ensure of process. Managed; Function clean.
            break
        except ValueError:
            print("ERROR! Enter a Number/Decimal Only. Try again.")        
    return temp

def get_Unit() -> str:
    """Ask for Temperature Unit (re-prompt when not F/C/K)"""
    while True:
        unit = str(input("Enter Unit: ").strip().upper())       
        while unit != 'F' and unit != 'C' and unit != 'K':
            unit = str(input("Must be 'F/C/K': ").strip().upper())
        break
    #print(unit) #Tester Flag Pass{X} --Notes: Checked the error loop of wrong letter, remove Len
    # this while covered it, no try/except since string validation
    return unit

def abs_Zero(temp: float, unit: str) -> bool:
    """Bool check input is within physical range"""
    if unit == 'K' and temp < ABS_ZERO_K:
        return False
    elif unit == 'K' and temp >= ABS_ZERO_K:
        return True
    elif unit == 'C' and temp < ABS_ZERO_C:
        return False
    elif unit == 'C' and temp >= ABS_ZERO_C:
        return True
    elif unit == 'F' and temp < ABS_ZERO_F:
        return False
    elif unit == 'F' and temp >= ABS_ZERO_F:
        return True
    
def to_Celsius(temp: float, unit: str) -> float:
    """Base conversion to celsius for temp intermission"""
    if unit == 'K':
        temp = temp - 273.15
        return temp
    elif unit == 'F':
        temp = (temp - 32) * 5.0 / 9.0
        return temp
    elif unit == 'C':
        temp = temp
        return temp

def celsFahr(celsius: float) -> float:
    """'cels'ius convert to 'Fahr'enheit for slick application"""
    celsius = celsius * 9.0 / 5.0 + 32.0
    return celsius

def celsKelv(celsius:float) -> float:
    """'cels'ius convert to 'Kelv'in for slick application"""
    celsius = celsius + 273.15
    return celsius

def disp_Res(orig_temp: float, unit: str):
    """'disp'lays 'Res'ults for user interpretation, all three conversions."""
    print(f"{orig_temp:,.2f}° {unit} converted is:\n")
    if unit == 'C':
        print(f"Kelvin: {celsKelv(orig_temp):,.2f}° K\n")
        print(f"Fahrenheit: {celsFahr(orig_temp):,.2f}° F")
    elif unit == 'F':
        orig_temp = to_Celsius(orig_temp, unit)
        print(f"Kelvin: {celsKelv(orig_temp):,.2f}° K\n")
        print(f"Celsius: {orig_temp:,.2f}° C")
    elif unit == 'K':
        orig_temp = to_Celsius(orig_temp, unit)
        print(f"Fahrenheit: {celsFahr(orig_temp):,.2f}° F\n")
        print(f"Celsius: {orig_temp:,.2f}° C")

def runAgain() -> bool:
    """Ask for Y/N only to restart program."""
    while True:
        ask = str(input("Would you like to convert a new number or leave? (Y/N): ").strip().upper())
        while ask != 'Y' and ask != 'N':
            ask = str(input("ERROR! Must be Y/N to Y: Redo; N: Leave; Choice: ").strip().upper())
        break
    if ask == 'Y':
        return True
    elif ask == 'N':
        return False
    
def main():
    """Main Function of the Program"""


#Need to Unit Test and Run:
#Input, Errors, Temps <F:{};C:{};K:{}>
#Run the loop, and make it happen for User Extension of Permission. {}
#Clean Up VSCode CoPilot advisory {}

File: cpp/tri_temp.cpp
//Date of Creation: 01/03/2026
// Takes User Input, asks for unit, then conversion happens.
// User is given all three units of conversion (all in a three display display) enter new/quit loop.
#include <iostream>
#include <iomanip>
#include <string>

using namespace std;
double tempNum()
{
    double num;
    cout << "Enter a Number for Temperature: ";
    cin >> num;
    //logic is, if anything is not a number, characters + string becomes 0;
    //cout << num << endl; //Tester Flag --Number showed, letters made it still 0; use for input check.
    
    return num;
}
char getUnit()
{
    //ask for character that is the unit of reference to convert.
    char unit;
    cout << "Enter the unit of Temperature (K/C/F): ";
    cin >> unit;
    //cout << unit << endl; //Tester Flag --worked
    unit = toupper(unit);
    //cout << unit << endl; //toUpper Tester Flag --worked
    return unit;
}
bool validUnit(char u)
{
    if (u == 'C' || u == 'K' || u == 'F')
    {
        return true;
    }
    else
    {
        return false;
    }
}
bool validTempConv(double t, char u)
{
    //if-chain for temperature validation of C, F, K, as Physics refuse these.
    
    if (u == 'K')
    {
        if (t < 0.0)
        {
            return false;
        }
    }
    else if (u == 'F')
        {
        if (t < -459.67)
        {
            return false;
        }
        }
    else if (u == 'C')
        {
        if (t < -273.15)
        {
            return false;
        }
            
        }
    return true;

}
double toCelsius(double t, char u)
{
    //if check to make temperature convert to Celsius as a medium
    
    if (u == 'F')
    {
        t = (t - 32) * 5.0 / 9.0;
        return t;
    }
    if (u == 'K')
    {
        t = t - 273.15;
        return t;
    }
    if (u == 'C')
    {
        t = t;
        return t;
    }
    return t;
}
double cToF(double c)
{
    //takes Celsius and makes it Fahrenheit for display.
    c = c * 9.0 / 5.0 + 32.0;
    return c;
}
double cToK(double c)
{
    c = c + 273.15;
    return c;
}
void dispRes(double t, char u)
{
    string banner = "=+=+=+=+=+=+=+=+=+=+=+=+=\n";
    cout << setprecision(2) << fixed << endl;
    cout << banner << "Results:\n" << banner;
    if (u == 'F')
    {
        cout << endl << "|" << setw(5) << left << t << " " << u << " becomes: " << endl;
        cout << "|Celsius: " << setw(5) << left << toCelsius(t, u) << " C" << endl;
        cout << "|Kelvin: " << setw(6) << left << cToK(toCelsius(t, u)) << " K" << endl;
    }
    if (u == 'C')
    {
        cout << endl << "|" << setw(5) << left << t << " " << u << " becomes: " << endl;
        cout << "|Fahrenheit: " << setw(5) << left << cToF(t) << " F" << endl;
        cout << "|Kelvin: " << setw(6) << left << cToK(t) << " K" << endl;
    }
    if (u == 'K')
    {
        cout << endl << "|" << setw(5) << left << t << " " << u << " becomes: " << endl;
        cout << "|Celsius: " << setw(5) << left << toCelsius(t, u) << " C" << endl;
        cout << "|Fahrenheit: " << setw(5) << left << cToF(toCelsius(t, u)) << " F" << endl;
    }
}
char replayChar()
{
    char resp;
    cout << "|Would you like to convert a new temp? (Y/N): ";
    cin >> resp;
    //cout << resp << endl; //tester flag-- got the closer to show, bool wasn't reading
    resp = toupper(resp);
    //cout << "Modified: " << resp << endl; //tester flag--
    return resp;
}
bool replayCheck(char r)
{
    //character check for the bool,
    if (r == 'Y' || r == 'N')
    {
        return true;
    }
    else
    {
        return false;
    }
}


int main()
{
    bool isRunning = true;
    while (isRunning)
    {
        double input = tempNum();
        double temp = (input);
        char baseU = getUnit();
        while (!validUnit(baseU))
        {
            cout << "|ERROR: Not a unit. Enter (K/C/F).\n";
            baseU = getUnit();
        }
        while (!validTempConv(input, baseU))
        {
            cout << "|ERROR: Below Physical Limits. Retry\n";
            input = tempNum();
        }
        
        dispRes(input, baseU);

        char r = replayChar();
        while (!replayCheck(r))
        {
            cout << "|Error: Please enter Y or N.\n";
            r = replayChar();
        }
        if (r == 'N')
        {
            system("cls");
            cout << "|Thank you for using Tri-Temp!" << endl;
            isRunning = false;
        }
        else if (r == 'Y')
        {
            isRunning = true;
            system("cls");
        }
    }
}


